test/micro/org/openjdk/bench/java/lang/ref/CleanerGC.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.bench.java.lang.ref;
+
+import java.lang.ref.Cleaner;
+import java.lang.ref.Cleaner.Cleanable;
+import java.util.LinkedList;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import org.openjdk.jmh.annotations.*;
+import org.openjdk.jmh.infra.Blackhole;
+
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@State(Scope.Thread)
+@BenchmarkMode(Mode.AverageTime)
+@Warmup(iterations = 3, time = 1, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 3, time = 1, timeUnit = TimeUnit.SECONDS)
+@Fork(value = 5, jvmArgs = {"-Xmx512m", "-Xms512m", "-XX:+AlwaysPreTouch"})
+public class CleanerGC {
+
+    @Param({"16384", "65536", "262144", "1048576", "4194304"})
+    int count;
+
+    // Deliberately a linked list to avoid exposing external parallelism to GC.
+    Target prev;
```

Comment:
Is `CleanerGC.prev` used?

Reply:
Yes. It maintains the root of a reachable linked list of `Target`-s.

Reply:
Doesn't it get covered by the local `prev` in `Setup()`? Should that be removed?

Reply:
No. Local `prev` goes out of scope when `setup()` finishes. We need something that keeps the list alive, like a benchmark object field, which `prev` is.

Reply:
Sorry for being dense, but where is a value being set to `CleanerGC.prev` ?

Reply:
Ah (facepalms), I see, I thought I assigned it! Damn. Let me fix and re-measure.

Reply:
Redone the benchmark. We actually want all targets to be reachable, and only expose the Cleaner-side linked lists to the test. New performance data is updated here: https://github.com/openjdk/jdk/pull/22043#issuecomment-2470928360

Reply:
Would this be considered a benchmark of the GC's PhantomReference management, then - deciding that all the `PhantomCleanerRef`s are still reachable, perhaps?

To measure `CleanableList`'s list manipulation at GC-time, a `Target` needs to become unreachable, so its `PhantomCleanerRef` will come off the `ReferenceQueue` (CleanerImp.java L140). Right?

Reply:
How about this:

> ...the problem is in GC time hiccups to walk the large linked lists.

Though, for this test, `CleanableList` itself is not being walked.

Rather, `CleanableList` itself is a factor in as much as it exists on the heap, populated with `PhantomCleanableRef`s.

So this test measures what effect, if any, the conversion from the old linked list to new CleanableList has on GC, when referents are _not_ becoming unreachable.

Reply:
> Though, for this test, CleanableList itself is not being walked.

I think this is a matter of definition of "CleanableList itself is not being walked". 

The Java code does not walk the list in this test, but GC does the walk the list as part of GC cycle. This is what this test sets up: it builds the large list of cleanable objects, makes sure the cleanable objects are reachable, so associated cleaners stay in the list, and then forces GC to mark the entire heap. When there is a large linked list formed by old `ClenableList` implementation, GC times suffer, because GC cannot parallelize marking work. This is why this benchmark improves with new implementation: GC becomes more parallel/efficient.

> So this test measures what effect, if any, the conversion from the old linked list to new CleanableList has on GC, when referents are not becoming unreachable.

Yes. And "if any" in this test shows up as 2..3x improvement in GC times :)

Reply:
Congratulations !

Reply:
So, @bchristi-git -- are you happy with current version, or?

Reply:
Sorry, U.S. holiday. Yes, I'm happy with the current version. Thanks for answering my questions.